File: packages/core-go/routing/extract_test.go
package routing

import (
	"reflect"
	"testing"

	"github.com/REDISHFISH/BLUEWHALE/packages/core-go/address"
)

const (
	testBaseG = "GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSI"
	testMuxed = "MAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQACABAAAAAAAAAAEVIG"
)

func TestExtractRouting_RoutingMatrix(t *testing.T) {
	tests := []struct {
		name     string
		input    RoutingInput
		expected RoutingResult
	}{
		{
			name: "g_address_without_memo_routes_none",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "none",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings:               []address.Warning{},
			},
		},
		{
			name: "memo-id",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "100",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("100"),
				RoutingSource:          "memo",
				Warnings:               []address.Warning{},
			},
		},
		{
			name: "memo-id-zero",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "0",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("0"),
				RoutingSource:          "memo",
				Warnings:               []address.Warning{},
			},
		},
		{
			name: "memo-id-max-uint64",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "18446744073709551615",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("18446744073709551615"),
				RoutingSource:          "memo",
				Warnings:               []address.Warning{},
			},
		},
		{
			name: "memo-id-empty",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings: []address.Warning{
					{
						Code:     address.WarnMemoIDInvalidFormat,
						Severity: "warn",
						Message:  "MEMO_ID was empty, non-numeric, or exceeded uint64 max.",
					},
				},
			},
		},
		{
			name: "memo-id-non-numeric",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "abc",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings: []address.Warning{
					{
						Code:     address.WarnMemoIDInvalidFormat,
						Severity: "warn",
						Message:  "MEMO_ID was empty, non-numeric, or exceeded uint64 max.",
					},
				},
			},
		},
		{
			name: "memo-id-overflow",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "18446744073709551616",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings: []address.Warning{
					{
						Code:     address.WarnMemoIDInvalidFormat,
						Severity: "warn",
						Message:  "MEMO_ID was empty, non-numeric, or exceeded uint64 max.",
					},
				},
			},
		},
		{
			name: "memo-id-normalization",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "id",
				MemoValue:   "007",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("7"),
				RoutingSource:          "memo",
				Warnings: []address.Warning{
					{
						Code:     address.WarnNonCanonicalRoutingID,
						Severity: "warn",
						Message:  "Memo routing ID had leading zeros. Normalized to canonical decimal.",
						Normalization: &address.Normalization{
							Original:   "007",
							Normalized: "7",
						},
					},
				},
			},
		},
		{
			name: "memo-text",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "text",
				MemoValue:   "200",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("200"),
				RoutingSource:          "memo",
				Warnings:               []address.Warning{},
			},
		},
		{
			name: "memo-hash",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "hash",
				MemoValue:   "not-a-routing-id",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings: []address.Warning{
					{
						Code:     address.WarnUnsupportedMemoType,
						Severity: "warn",
						Message:  "Memo type hash is not supported for routing.",
						Context: &address.WarningContext{
							MemoType: "hash",
						},
					},
				},
			},
		},
		{
			name: "memo-return",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "return",
				MemoValue:   "also-not-a-routing-id",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings: []address.Warning{
					{
						Code:     address.WarnUnsupportedMemoType,
						Severity: "warn",
						Message:  "Memo type return is not supported for routing.",
						Context: &address.WarningContext{
							MemoType: "return",
						},
					},
				},
			},
		},
		{
			name: "g_address_with_unknown_memo_type_warns_unknown",
			input: RoutingInput{
				Destination: testBaseG,
				MemoType:    "memo_blob",
				MemoValue:   "opaque",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              nil,
				RoutingSource:          "none",
				Warnings: []address.Warning{
					{
						Code:     address.WarnUnsupportedMemoType,
						Severity: "warn",
						Message:  "Unrecognized memo type: memo_blob",
						Context: &address.WarningContext{
							MemoType: "unknown",
						},
					},
				},
			},
		},
		{
			name: "muxed",
			input: RoutingInput{
				Destination: testMuxed,
				MemoType:    "none",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("9007199254740993"),
				RoutingSource:          "muxed",
				Warnings:               []address.Warning{},
			},
		},
		{
			name: "m_address_with_routing_memo_warns_memo_present_with_muxed",
			input: RoutingInput{
				Destination: testMuxed,
				MemoType:    "id",
				MemoValue:   "42",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("9007199254740993"),
				RoutingSource:          "muxed",
				Warnings: []address.Warning{
					{
						Code:     address.WarnMemoPresentWithMuxed,
						Severity: "warn",
						Message:  "Routing ID found in both M-address and Memo. M-address ID takes precedence.",
					},
				},
			},
		},
		{
			name: "m_address_with_non_routing_memo_warns_memo_ignored",
			input: RoutingInput{
				Destination: testMuxed,
				MemoType:    "text",
				MemoValue:   "not-a-routing-id",
			},
			expected: RoutingResult{
				DestinationBaseAccount: testBaseG,
				RoutingID:              NewRoutingID("9007199254740993"),
				RoutingSource:          "muxed",
				Warnings: []address.Warning{
					{
						Code:     address.WarnMemoIgnoredForMuxed,
						Severity: "info",
						Message:  "Memo present with M-address. Any potential routing ID in memo is ignored.",
					},
				},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assertRoutingResult(t, ExtractRouting(tt.input), tt.expected)
		})
	}
}

func TestExtractRouting_ContractSourceClearsRoutingState(t *testing.T) {
	t.Run("contract-source", func(t *testing.T) {
		contractAddress, err := address.EncodeStrKey(address.VersionByteC, make([]byte, 32))
		if err != nil {
			t.Fatalf("failed to generate contract address: %v", err)
		}

		result := ExtractRouting(RoutingInput{
			Destination:   testBaseG,
			MemoType:      "id",
			MemoValue:     "100",
			SourceAccount: contractAddress,
		})

		expected := RoutingResult{
			RoutingSource: "none",
			Warnings: []address.Warning{
				{
					Code:     address.WarnContractSenderDetected,
					Severity: "info",
					Message:  "Contract source detected. Routing state cleared.",
				},
			},
		}

		assertRoutingResult(t, result, expected)
	})
}

func assertRoutingResult(t *testing.T, got, want RoutingResult) {
	t.Helper()

	if got.DestinationBaseAccount != want.DestinationBaseAccount {
		t.Errorf("DestinationBaseAccount = %v, want %v", got.DestinationBaseAccount, want.DestinationBaseAccount)
	}
	if !routingIDEqual(got.RoutingID, want.RoutingID) {
		t.Errorf("RoutingID = %v, want %v", got.RoutingID.String(), want.RoutingID.String())
	}
	if got.RoutingSource != want.RoutingSource {
		t.Errorf("RoutingSource = %v, want %v", got.RoutingSource, want.RoutingSource)
	}
	if !reflect.DeepEqual(got.Warnings, want.Warnings) {
		t.Errorf("Warnings = %#+v, want %#+v", got.Warnings, want.Warnings)
	}
	if !reflect.DeepEqual(got.DestinationError, want.DestinationError) {
		t.Errorf("DestinationError = %#v, want %#v", got.DestinationError, want.DestinationError)
	}
}

func routingIDEqual(a, b *RoutingID) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return a.String() == b.String()
}

func TestNormalizeUnsupportedMemoType(t *testing.T) {
	t.Parallel()

	testCases := map[string]string{
		"hash":             "hash",
		"return":           "return",
		"none":             "",
		"id":               "",
		"text":             "",
		"":                 "",
		"MEMO_HASH":        "hash",
		"memo-return":      "return",
		"Memo_Return":      "return",
		"memohash":         "hash",
		"m_e_m_o_h_a_s_h":  "hash",
		"HASH":             "",
		"memo_hash_extra":  "",
		"memo_returnx":     "",
		"something-longer": "",
		"___":              "",
	}

	for input, want := range testCases {
		if got := normalizeUnsupportedMemoType(input); got != want {
			t.Errorf("normalizeUnsupportedMemoType(%q) = %q, want %q", input, got, want)
		}
	}
}

func TestNormalizeUnsupportedMemoTypeDoesNotAllocate(t *testing.T) {
	for _, input := range []string{"hash", "return", "none", "id", "text", "MEMO_HASH", "memo-return", "unknown_type_value"} {
		allocs := testing.AllocsPerRun(100, func() {
			_ = normalizeUnsupportedMemoType(input)
		})
		if allocs != 0 {
			t.Errorf("normalizeUnsupportedMemoType(%q) allocs = %v, want 0", input, allocs)
		}
	}
}

File: packages/core-ts/src/test/extract.test.ts
/**
 * extractRouting() – structured Warning diagnostics test suite
 *
 * The Warning interface emits typed objects ({ code, severity, message })
 * rather than plain strings. This suite validates that each warning branch
 * in extract.ts injects the correct schema.
 *
 * Test categories
 * ───────────────
 *  1. MEMO_PRESENT_WITH_MUXED   – M-address + numeric memo-id conflict
 *  2. MEMO_IGNORED_FOR_MUXED    – M-address + non-routable memo type
 *  3. CONTRACT_SENDER_DETECTED  – C-address is rejected at the routing guard
 *  4. MEMO_ID_INVALID_FORMAT    – G-address + empty / non-numeric memo-id
 *  5. MEMO_TEXT_UNROUTABLE      – G-address + unsupported memo type (hash)
 *  6. NON_CANONICAL_ROUTING_ID  – G-address + leading-zero memo-id
 *  7. Multi-warning              – leading-zero memo-id that exceeds uint64 max
 *                                  triggers NON_CANONICAL_ROUTING_ID +
 *                                  MEMO_ID_INVALID_FORMAT simultaneously
 */

import { describe, it, expect, beforeEach } from "vitest";
import { encodeMuxed } from "../muxed/encode";
import { extractRouting, ExtractRoutingError } from "../routing/extract";
import type { RoutingInput, RoutingResult, Warning } from "../routing/types";

// ─── Test fixtures ────────────────────────────────────────────────────────────

// Well-known G-address from spec/vectors.json (muxed_encode cases).
const G_ADDRESS = "GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSI";
const ROUTING_ID = 42n;
const M_ADDRESS = encodeMuxed(G_ADDRESS, ROUTING_ID);

// Helper: build a complete RoutingInput with sensible defaults.
function input(
  destination: string,
  memoType = "none",
  memoValue: string | null = null
): RoutingInput {
  return { destination, memoType, memoValue, sourceAccount: null };
}

// ─── 1. MEMO_PRESENT_WITH_MUXED ───────────────────────────────────────────────

describe("MEMO_PRESENT_WITH_MUXED warning", () => {
  let result: RoutingResult;

  // Trigger: M-address destination + numeric memo-id supplied simultaneously.
  beforeEach(() => {
    result = extractRouting(input(M_ADDRESS, "id", "99999"));
  });

  it("emits exactly one warning", () => {
    expect(result.warnings).toHaveLength(1);
  });

  it("warning.code is MEMO_PRESENT_WITH_MUXED", () => {
    expect(result.warnings[0].code).toBe("MEMO_PRESENT_WITH_MUXED");
  });

  it("warning.severity is 'warn'", () => {
    expect(result.warnings[0].severity).toBe("warn");
  });

  it("warning.message is a non-empty string", () => {
    expect(typeof result.warnings[0].message).toBe("string");
    expect(result.warnings[0].message.length).toBeGreaterThan(0);
  });

  it("routing still resolves from the M-address (source is 'muxed')", () => {
    expect(result.routingSource).toBe("muxed");
    expect(result.destinationBaseAccount).toBe(G_ADDRESS);
  });
});

// ─── 2. MEMO_IGNORED_FOR_MUXED ────────────────────────────────────────────────

describe("MEMO_IGNORED_FOR_MUXED warning", () => {
  let result: RoutingResult;

  // Trigger: M-address destination + a non-numeric text memo.
  beforeEach(() => {
    result = extractRouting(input(M_ADDRESS, "text", "order-ref-abc"));
  });

  it("emits exactly one warning", () => {
    expect(result.warnings).toHaveLength(1);
  });

  it("warning.code is MEMO_IGNORED_FOR_MUXED", () => {
    expect(result.warnings[0].code).toBe("MEMO_IGNORED_FOR_MUXED");
  });

  it("warning.severity is 'info'", () => {
    expect(result.warnings[0].severity).toBe("info");
  });

  it("warning.message is a non-empty string", () => {
    expect(typeof result.warnings[0].message).toBe("string");
    expect(result.warnings[0].message.length).toBeGreaterThan(0);
  });

  it("routing ID still comes from the M-address", () => {
    expect(result.routingSource).toBe("muxed");
    expect(result.destinationBaseAccount).toBe(G_ADDRESS);
  });
});

// ─── 3. C-address zero-throw policy ──────────────────────────────────────────
//
// C-addresses used to throw ExtractRoutingError but the zero-throw policy
// (Issue #77) means extractRouting now returns a structured result with an
// INVALID_DESTINATION warning instead of throwing.

describe("C-address – zero-throw policy (#77)", () => {
  // A well-formed Stellar contract address (C-prefix).
  const C_ADDRESS = "CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC";

  it("does NOT throw for a contract address", () => {
    expect(() => extractRouting(input(C_ADDRESS))).not.toThrow();
  });

  it("returns a structured result with routingSource 'none'", () => {
    const result = extractRouting(input(C_ADDRESS));
    expect(result.routingSource).toBe("none");
    expect(result.destinationBaseAccount).toBeNull();
    expect(result.routingId).toBeNull();
  });

  it("emits an INVALID_DESTINATION error-severity warning", () => {
    const result = extractRouting(input(C_ADDRESS));
    const warning = result.warnings.find((w) => w.code === "INVALID_DESTINATION");
    expect(warning).toBeDefined();
    expect(warning!.severity).toBe("error");
  });
});

// ─── 3b. CONTRACT_SENDER_DETECTED (contract source account) ──────────────────
//
// When the *sender* is a Soroban contract, routing state is cleared and a
// CONTRACT_SENDER_DETECTED info warning is returned (parity with Go and Dart).

describe("CONTRACT_SENDER_DETECTED – contract source account", () => {
  const C_SOURCE = "CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC";
  const expected: RoutingResult = {
    destinationBaseAccount: null,
    routingId: null,
    routingSource: "none",
    warnings: [
      {
        code: "CONTRACT_SENDER_DETECTED",
        severity: "info",
        message: "Contract source detected. Routing state cleared.",
      },
    ],
  };

  it("clears routing state for a G destination with a memo", () => {
    const result = extractRouting({
      ...input(G_ADDRESS, "id", "100"),
      sourceAccount: C_SOURCE,
    });
    expect(result).toEqual(expected);
  });

  it("clears routing state for an M destination", () => {
    const result = extractRouting({
      ...input(M_ADDRESS),
      sourceAccount: C_SOURCE,
    });
    expect(result).toEqual(expected);
  });

  it("does not trigger for a G source account", () => {
    const result = extractRouting({
      ...input(G_ADDRESS, "id", "100"),
      sourceAccount: G_ADDRESS,
    });
    expect(result.routingSource).toBe("memo");
    expect(result.routingId).toBe("100");
  });

  it("ignores an unparseable source account", () => {
    const result = extractRouting({
      ...input(G_ADDRESS, "id", "100"),
      sourceAccount: "CNOTAVALIDCONTRACT",
    });
    expect(result.routingSource).toBe("memo");
  });
});

// ─── 4. MEMO_ID_INVALID_FORMAT ────────────────────────────────────────────────

describe("MEMO_ID_INVALID_FORMAT warning", () => {
  it("emits MEMO_ID_INVALID_FORMAT for an empty memo-id value", () => {
    const result = extractRouting(input(G_ADDRESS, "id", ""));
    const codes = result.warnings.map((w: Warning) => w.code);
    expect(codes).toContain("MEMO_ID_INVALID_FORMAT");
  });

  it("warning object has severity 'warn'", () => {
    const result = extractRouting(input(G_ADDRESS, "id", ""));
    const warning = result.warnings.find(
      (w: Warning) => w.code === "MEMO_ID_INVALID_FORMAT"
    );
    expect(warning).toBeDefined();
    expect(warning!.severity).toBe("warn");
    expect(typeof warning!.message).toBe("string");
  });

  it("emits MEMO_ID_INVALID_FORMAT for a non-numeric memo-id", () => {
    const result = extractRouting(input(G_ADDRESS, "id", "not-a-number"));
    const codes = result.warnings.map((w: Warning) => w.code);
    expect(codes).toContain("MEMO_ID_INVALID_FORMAT");
  });

  it("routing source falls back to 'none' when memo-id is invalid", () => {
    const result = extractRouting(input(G_ADDRESS, "id", ""));
    expect(result.routingSource).toBe("none");
    expect(result.routingId).toBeNull();
  });
});

// ─── 5. UNSUPPORTED_MEMO_TYPE / MEMO_TEXT_UNROUTABLE ──────────────────────────

describe("UNSUPPORTED_MEMO_TYPE warning", () => {
  // A memo type that cannot carry a routing ID is reported as
  // UNSUPPORTED_MEMO_TYPE with the `context.memoType` that spec/schema.json
  // requires. MEMO_TEXT_UNROUTABLE is reserved for an actual MEMO_TEXT value
  // that is not a numeric uint64.
  it("emits UNSUPPORTED_MEMO_TYPE for memoType 'hash'", () => {
    const result = extractRouting(input(G_ADDRESS, "hash", "abc123"));
    const codes = result.warnings.map((w: Warning) => w.code);
    expect(codes).toContain("UNSUPPORTED_MEMO_TYPE");
    expect(codes).not.toContain("MEMO_TEXT_UNROUTABLE");
  });

  it("emits UNSUPPORTED_MEMO_TYPE for memoType 'return'", () => {
    const result = extractRouting(input(G_ADDRESS, "return", "abc123"));
    const codes = result.warnings.map((w: Warning) => w.code);
    expect(codes).toContain("UNSUPPORTED_MEMO_TYPE");
    expect(codes).not.toContain("MEMO_TEXT_UNROUTABLE");
  });

  it("emits UNSUPPORTED_MEMO_TYPE with context.memoType 'unknown' for an unrecognized memo type", () => {
    const result = extractRouting(input(G_ADDRESS, "signed", "abc123"));
    expect(result.warnings).toContainEqual({
      code: "UNSUPPORTED_MEMO_TYPE",
      severity: "warn",
      message: "Unrecognized memo type: signed",
      context: { memoType: "unknown" },
    });
  });

  it("warning object has severity 'warn', a non-empty message and a context", () => {
    const result = extractRouting(input(G_ADDRESS, "hash", "abc123"));
    const warning = result.warnings.find(
      (w: Warning) => w.code === "UNSUPPORTED_MEMO_TYPE"
    );
    expect(warning).toBeDefined();
    expect(warning!.severity).toBe("warn");
    expect(warning!.message.length).toBeGreaterThan(0);
    expect((warning as { context: { memoType: string } }).context).toEqual({
      memoType: "hash",
    });
  });
});

describe("MEMO_TEXT_UNROUTABLE warning", () => {
  it("emits MEMO_TEXT_UNROUTABLE for a non-numeric MEMO_TEXT", () => {
    const result = extractRouting(input(G_ADDRESS, "text", "not-a-number"));
    const codes = result.warnings.map((w: Warning) => w.code);
    expect(codes).toContain("MEMO_TEXT_UNROUTABLE");
    expect(codes).not.toContain("UNSUPPORTED_MEMO_TYPE");
  });

  it("warning object has severity 'warn' and a non-empty message", () => {
    const result = extractRouting(input(G_ADDRESS, "text", "not-a-number"));
    const warning = result.warnings.find(
      (w: Warning) => w.code === "MEMO_TEXT_UNROUTABLE"
    );
    expect(warning).toBeDefined();
    expect(warning!.severity).toBe("warn");
    expect(warning!.message.length).toBeGreaterThan(0);
  });
});

// ─── 6. NON_CANONICAL_ROUTING_ID ─────────────────────────────────────────────

describe("NON_CANONICAL_ROUTING_ID warning", () => {
  // Trigger: G-address + memoType "id" with leading zeros (e.g. "007").
  let result: RoutingResult;

  beforeEach(() => {
    result = extractRouting(input(G_ADDRESS, "id", "007"));
  });

  it("emits exactly one warning", () => {
    expect(result.warnings).toHaveLength(1);
  });

  it("warning.code is NON_CANONICAL_ROUTING_ID", () => {
    expect(result.warnings[0].code).toBe("NON_CANONICAL_ROUTING_ID");
  });

  it("warning.severity is 'warn'", () => {
    expect(result.warnings[0].severity).toBe("warn");
  });

  it("warning carries normalization metadata with original and normalized values", () => {
    const w = result.warnings[0] as Extract<Warning, { normalization: unknown }>;
    expect(w.normalization.original).toBe("007");
    expect(w.normalization.normalized).toBe("7");
  });

  it("routingId is normalized to the canonical decimal form", () => {
    expect(result.routingId).toBe("7");
    expect(result.routingSource).toBe("memo");
  });
});

// ─── 7. Multi-warning: NON_CANONICAL_ROUTING_ID + MEMO_ID_INVALID_FORMAT ─────
//
// A memo-id value with leading zeros that also exceeds UINT64_MAX triggers
// TWO warnings at once:
//   1. NON_CANONICAL_ROUTING_ID  (from normalizeMemoTextId stripping leading zero)
//   2. MEMO_ID_INVALID_FORMAT    (because the stripped value exceeds uint64 max)

describe("multi-warning: NON_CANONICAL_ROUTING_ID + MEMO_ID_INVALID_FORMAT", () => {
  // "018446744073709551616" = leading zero + value one above UINT64_MAX
  const OVERFLOW_WITH_LEADING_ZERO = "018446744073709551616";
  let result: RoutingResult;

  beforeEach(() => {
    result = extractRouting(input(G_ADDRESS, "id", OVERFLOW_WITH_LEADING_ZERO));
  });

  it("emits exactly two warnings", () => {
    expect(result.warnings).toHaveLength(2);
  });

  it("first warning is NON_CANONICAL_ROUTING_ID", () => {
    expect(result.warnings[0].code).toBe("NON_CANONICAL_ROUTING_ID");
  });

  it("second warning is MEMO_ID_INVALID_FORMAT", () => {
    expect(result.warnings[1].code).toBe("MEMO_ID_INVALID_FORMAT");
  });

  it("both warnings have severity 'warn'", () => {
    for (const w of result.warnings) {
      expect(w.severity).toBe("warn");
    }
  });

  it("NON_CANONICAL_ROUTING_ID normalization metadata is correct", () => {
    const w = result.warnings[0] as Extract<Warning, { normalization: unknown }>;
    expect(w.normalization.original).toBe(OVERFLOW_WITH_LEADING_ZERO);
    expect(w.normalization.normalized).toBe("18446744073709551616");
  });

  it("routing falls back to 'none' because the value exceeds uint64 max", () => {
    expect(result.routingSource).toBe("none");
    expect(result.routingId).toBeNull();
  });

  it("all warning objects carry a non-empty message string", () => {
    for (const w of result.warnings) {
      expect(typeof w.message).toBe("string");
      expect(w.message.length).toBeGreaterThan(0);
    }
  });
});
